README.md
# NAME

DataDog::DogStatsd - A Perl client for DogStatsd

# SYNOPSIS

        use DataDog::DogStatsd;

        my $statsd = DataDog::DogStatsd->new;
        $statsd->increment( 'user.login', { tags => [ $s->cs->param( 'loginName' ) ] } );
        $statsd->timing('page.load', 320);
        $statsd->gauge('users.online', 100);

        # namespace version
        my $statsd = DataDog::DogStatsd->new(namespace => 'account.'); # note the '.' at the ending
        $statsd->increment('activate'); # actually as account.activate

# DESCRIPTION

Statsd: A DogStatsd client (https://www.datadoghq.com)

The difference between [Net::Statsd](https://metacpan.org/pod/Net%3A%3AStatsd) and this is that it supports tags and namespace.

# METHODS

## new

- host

    default to 127.0.0.1

- port

    default to 8125

- namespace

    default to ''

## increment

        $statsd->increment( 'page.views' );
        $statsd->increment( 'test.stats', { sample_rate => 0.5 } );
        $statsd->increment( 'user.login', { tags => [ $s->cs->param( 'loginName' ) ] } );

Sends an increment (count = 1) for the given stat to the statsd server.

## decrement

        $statsd->decrement( 'products.available' );

Sends a decrement (count = -1) for the given stat to the statsd server.

## count

        $statsd->count( 'test.stats', 2 );

Sends an arbitrary count for the given stat to the statsd server.

## gauge

        $statsd->gauge('users.online', 100);
        $statsd->gauge('users.online', 100, { tags => ['tag1', 'tag2'] });

Sends an arbitrary gauge value for the given stat to the statsd server.

This is useful for recording things like available disk space, memory usage, and the like, which have different semantics than counters.

## histogram

        $statsd->histogram($stats, $value);

Sends a value to be tracked as a histogram to the statsd server.

## timing

        $statsd->timing('page.load', 320);
        $statsd->timing('page.load', 320, { sample_rate => 0.5, tags => ['tag1', 'tag2'] });

Sends a timing (in ms) for the given stat to the statsd server. The sample\_rate determines what percentage of the time this report is sent. The statsd server then uses the sample\_rate to correctly track the average timing for the stat.

## set

        $statsd->set('visitors.uniques', $user_id);

Sends a value to be tracked as a set to the statsd server.

## event

        $statsd->event('event title', 'event text', { tags => ['tag1, 'tag2'] });
        $statsd->event('event title', 'event text', { tags => ['tag1', 'tag2'], alert_type => 'error'});

Sends an event indicating a specific incident occurring.

Use alert\_type option to specify the kind of event, available options are error, warning, info, and success.

# AUTHORS

Stefan Goethals <stefan@zipkid.eu>


Changes
Revision history for DataDog-DogStatsd

{{$NEXT}}

0.07      2022-10-25 06:45:09+00:00 UTC
      - Apply perlcritic

0.06  2019-12-04 00:00:00
        - Allow setting the DataDog host, port and namespace via DATADOG_AGENT_HOST,
          DATADOG_AGENT_PORT, and DATADOG_AGENT_NAMESPACE environment variables
        - Add stats_event

0.05  2018-07-02 15:54:34 +08
        - Add stats_histogram

0.04  2015-01-13 18:09:20 CST
        - sprintf %d for timing
        - more tests

0.03  2015-01-12 18:34:29 CST
        - add Helper

0.01  2015-01-12 18:30:33 CST
        - Initial release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.024.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 7.64;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A Perl client for DogStatsd",
  "AUTHOR" => "binary.com <BINARY\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "7.64"
  },
  "DISTNAME" => "DataDog-DogStatsd",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "DataDog::DogStatsd",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "IO::Socket::INET" => 0,
    "Time::HiRes" => 0,
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Basename" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IO::Select" => 0,
    "IPC::Open3" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::More" => "0.96",
    "lib" => 0,
    "vars" => 0
  },
  "VERSION" => "0.08",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IO::Select" => 0,
  "IO::Socket::INET" => 0,
  "IPC::Open3" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::More" => "0.96",
  "Time::HiRes" => 0,
  "base" => 0,
  "lib" => 0,
  "strict" => 0,
  "vars" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/DataDog/DogStatsd/Helper.pm
package DataDog::DogStatsd::Helper;

use strict;
use warnings;
our $VERSION = '0.06';

use base qw( Exporter );
our @EXPORT_OK = qw/stats_inc stats_dec stats_timing stats_gauge stats_count stats_histogram stats_event stats_timed/;

use DataDog::DogStatsd;
use Time::HiRes qw(tv_interval gettimeofday);

## no critic (Subroutines::RequireFinalReturn)
sub stats_inc {
    my @args = @_;
    # support stats_inc('test.blabla', 0.1); as well
    if (@args > 1 and $args[1] =~ /^[\d\.]+$/) {
        # actually ppl wants stats_count
        warn "stats_inc sample_rate makes no sense for more than 1\n" if $args[1] > 1;
        $args[1] = {sample_rate => $args[1]};
    }
    get_dogstatsd()->increment(@args);
}
sub stats_dec { get_dogstatsd()->decrement(@_); }

sub stats_timing {
    my @args = @_;
    # support stats_timing('connection_time', 1000 * $interval, 0.1); as well
    if (@args > 2 and $args[2] =~ /^[\d\.]+$/) {
        # actually ppl wants stats_count
        warn "stats_timing sample_rate makes no sense for more than 1\n" if $args[2] > 1;
        $args[2] = {sample_rate => $args[2]};
    }
    get_dogstatsd()->timing(@args);
}

sub stats_timed (&@) {    ## no critic (ProhibitSubroutinePrototypes, Subroutines::RequireArgUnpacking)
    my ($sub, $what, $opts) = splice @_, 0, 3;

    my $start = [gettimeofday];

    my (@res, $no_ex);
    if (wantarray) {
        $no_ex = eval { @res = $sub->(@_); 1; };
    } elsif (defined wantarray) {
        $no_ex = eval { $res[0] = $sub->(@_); 1; };
    } else {
        $no_ex = eval { $sub->(@_); 1; };
    }

    if ($no_ex) {
        get_dogstatsd()->timing($what, 1000 * tv_interval($start), $opts);
    } else {
        get_dogstatsd()->increment("$what.failure", $opts);
        die $@;    ## no critic (Variables::ProhibitEvilVariables)
    }

    return wantarray ? @res : $res[0];
}

sub stats_gauge     { get_dogstatsd()->gauge(@_); }
sub stats_count     { get_dogstatsd()->count(@_); }
sub stats_histogram { get_dogstatsd()->histogram(@_); }
sub stats_event     { get_dogstatsd()->event(@_); }

my $DOGSTATSD;

sub get_dogstatsd {
    $DOGSTATSD ||= DataDog::DogStatsd->new(
        host      => $ENV{DATADOG_AGENT_HOST},
        port      => $ENV{DATADOG_AGENT_PORT},
        namespace => $ENV{DATADOG_AGENT_NAMESPACE});
    return $DOGSTATSD;
}

sub set_dogstatsd {
    $DOGSTATSD = pop;
}

1;
__END__

=encoding utf-8

=head1 NAME

DataDog::DogStatsd::Helper - helper for L<DataDog::DogStatsd>

=head1 SYNOPSIS

    use DataDog::DogStatsd::Helper qw(stats_inc stats_dec stats_timing
                                      stats_gauge stats_count
                                      stats_histogram stats_timed);

    stats_inc('logins'); # shortcut for DataDog::DogStatsd->new->increment('logins')
    stats_dec('logins'); # shortcut for DataDog::DogStatsd->new->decrement('logins')
    stats_timing('test.timing', 1); # ->timing
    stats_gauge('test.gauge', 10); # ->gauge
    stats_count('test.count', 20); # ->count
    stats_histogram('test.histogram', 100); # ->histogram
    stats_event('event title', 'event text'); # ->event

    stats_timed {
        my @param = @_;
        # some code
        ...
    } 'test.timing', {tags => \@tags}, @params;

    DataDog::DogStatsd::Helper->get_dogstatsd;
    DataDog::DogStatsd::Helper->set_dogstatsd(DataDog::DogStatsd->new(...));

=head1 DESCRIPTION

DataDog::DogStatsd::Helper is a helper for L<DataDog::DogStatsd>. It's main purpose
is to maintain a global L<DataDog::DogStatsd> object.

=head1 FUNCTIONS

For all the functions below, C<$what> is the name of a metric sent to datadog. C<\%opts>
is a hash reference which can contain a key C<tags> pointing to an array of tags sent
to datadog along with the metric and a key C<sample_rate> with a number between C<0> and
C<1> as value. See
L<https://docs.datadoghq.com/developers/metrics/dogstatsd_metrics_submission/#sample-rates>
for more information on sample rates.

=head2 C<< stats_inc $what, \%opts >>

equivalent to C<< $dogstatsd->increment($what, \%opts) >>

=head2 C<< stats_inc $what, $sample_rate >>

equivalent to C<< $dogstatsd->increment($what, +{sample_rate=>$sample_rate}) >>

=head2 C<< stats_dec $what, \%opts >>

equivalent to C<< $dogstatsd->decrement($what, \%opts) >>

=head2 C<< stats_timing $what $millisec, \%opts >>

equivalent to C<< $dogstatsd->timing($what, $millisec, \%opts) >>

=head2 C<< stats_timing $what $millisec, $sample_rate >>

equivalent to C<< $dogstatsd->timing($what, $millisec, +{sample_rate=>$sample_rate}) >>

Note, C<$millisec> will be truncated to an integer value.

=head2 C<< stats_gauge $what, $measured_value, \%opts >>

equivalent to C<< $dogstatsd->gauge($what, $measured_value, \%opts) >>

=head2 C<< stats_count $what, $increment, \%opts >>

equivalent to C<< $dogstatsd->count($what, $increment, \%opts) >>

=head2 C<< stats_histogram $what, $measured_value, \%opts >>

equivalent to C<< $dogstatsd->histogram($what, $measured_value, \%opts) >>

=head2 C<< stats_event $title, $text, \%opts >>

equivalent to C<< $dogstatsd->event($what, $title, $text, \%opts) >>

Naturally, C<sample_rate> does not make sense in C<\%opts>. C<$opts{tags}>
is processed as usual. Besides that, the C<%opts> hash can contain the
following optional keys:

=over 4

=item date_happened

=item hostname

=item aggregation_key

=item priority

=item source_type_name

=item alert_type

=back

For more information see
L<https://docs.datadoghq.com/developers/events/dogstatsd/>

=head2 C<< stats_timed {BLOCK} $what, \%opts, @other_params >>

this offers a somewhat more convenient interface to timing a
piece of code. It's supposed to be called like this:

 my $value = stats_timed {
     # timed piece of code
     my ($p1, $p2, ...) = @_;
     ...
 } $what, {tags => \@tags}, $param1, $param2, ...;

Before the code block is executed the current time is taken with
C<Time::HiRes::gettimeofday>. Then the code block is executed
in an C<eval> environment. If no exception is thrown the time
is reported the same way C<stats_timing> does. If the code block
ends in an exception, the C<$what.failure> metric is incremented
passing the same tags.

If called in scalar context, the code block is also called in
scalar context and the resulting value is returned.

If called in list context, the code block is also called in
list context and the resulting list is returned.

If called in void context, the code block is also called in
void context.

=head2 C<< DataDog::DogStatsd::Helper->get_dogstatsd >>

returns the internal C<DataDog::DogStatsd> object. This can be used to
modify certain parameters like the C<namespace>.

This function is not exported. It can be called both as function and
as class method.

=head2 C<< DataDog::DogStatsd::Helper->set_dogstatsd(DataDog::DogStatsd->new) >>

allows to set the internal C<DataDog::DogStatsd> object. Useful mainly for
testing.

This function is not exported. It can be called both as function and
as class method.

=head1 AUTHOR

Fayland Lam E<lt>fayland@binary.comE<gt>

=head1 COPYRIGHT

Copyright 2015- Fayland Lam

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=cut


lib/DataDog/DogStatsd.pm
package DataDog::DogStatsd;

# ABSTRACT: A Perl client for DogStatsd

use strict;
use warnings;

our $VERSION = '0.08';

use IO::Socket::INET;

my %OPTS_KEYS = (
    date_happened    => 'd',
    hostname         => 'h',
    aggregation_key  => 'k',
    priority         => 'p',
    source_type_name => 's',
    alert_type       => 't'
);

sub new {    ## no critic (Subroutines::RequireArgUnpacking)
    my $classname = shift;
    my $class     = ref($classname) || $classname;
    my %p         = @_ % 2 ? %{$_[0]} : @_;

    $p{host}      ||= '127.0.0.1';
    $p{port}      ||= 8125;
    $p{namespace} ||= '';

    return bless \%p, $class;
}

sub _socket {
    my $self = shift;
    return $self->{_socket} if $self->{_socket};
    $self->{_socket} = IO::Socket::INET->new(
        PeerAddr => $self->{'host'},
        PeerPort => $self->{'port'},
        Proto    => 'udp'
    );
    return $self->{_socket};
}

sub namespace {
    my $self = shift;
    return $self->{'namespace'} unless @_;
    $self->{'namespace'} = shift;
}

sub increment {
    my $self = shift;
    my $stat = shift;
    my $opts = shift || {};
    $self->count($stat, 1, $opts);
}

sub decrement {
    my $self = shift;
    my $stat = shift;
    my $opts = shift || {};
    $self->count($stat, -1, $opts);
}

sub count {
    my $self  = shift;
    my $stat  = shift;
    my $count = shift;
    my $opts  = shift || {};
    $self->send_stats($stat, $count, 'c', $opts);
}

sub gauge {
    my $self  = shift;
    my $stat  = shift;
    my $value = shift;
    my $opts  = shift || {};
    $self->send_stats($stat, $value, 'g', $opts);
}

sub histogram {
    my $self  = shift;
    my $stat  = shift;
    my $value = shift;
    my $opts  = shift || {};
    $self->send_stats($stat, $value, 'h', $opts);
}

sub timing {
    my $self = shift;
    my $stat = shift;
    my $ms   = shift;
    my $opts = shift || {};
    $self->send_stats($stat, sprintf("%d", $ms), 'ms', $opts);
}

# Reports execution time of the provided block using {#timing}.
#
# @param [String] stat stat name
# @param [Hash] opts the options to create the metric with
# @option opts [Numeric] :sample_rate sample rate, 1 for always
# @option opts [Array<String>] :tags An array of tags
# @yield The operation to be timed
# @see #timing
# @example Report the time (in ms) taken to activate an account
#   $statsd.time('account.activate') { @account.activate! }
##def time(stat, opts={})
##	start = Time.now
##	result = yield
##	timing(stat, ((Time.now - start) * 1000).round, opts)
##	result
##end

sub set {
    my $self  = shift;
    my $stat  = shift;
    my $value = shift;
    my $opts  = shift || {};
    $self->send_stats($stat, $value, 's', $opts);
}

sub event {
    my $self  = shift;
    my $title = shift;
    my $text  = shift;
    my $opts  = shift || {};

    my $event_string = format_event($title, $text, $opts);

    $self->send_to_socket($event_string);
}

sub format_event {
    my $title = shift;
    my $text  = shift;
    my $opts  = shift || {};

    my $tags = delete $opts->{tags};

    $title =~ s/\n/\\n/g;
    $text  =~ s/\n/\\n/g;

    my $event_string_data = sprintf "_e{%d,%d}:%s|%s", length($title), length($text), $title, $text;

    for my $opt (keys %$opts) {
        if (my $key = $OPTS_KEYS{$opt}) {
            $opts->{$opt} =~ s/|//g;
            $event_string_data .= "|$key:$opts->{$opt}";
        }
    }

    if ($tags) {
        $event_string_data .= "|#" . join ",", map { s/|//g; $_ } @$tags;    ## no critic (ControlStructures::ProhibitMutatingListFunctions)
    }

    return $event_string_data;
}

sub send_stats {
    my $self  = shift;
    my $stat  = shift;
    my $delta = shift;
    my $type  = shift;
    my $opts  = shift || {};

    my $sample_rate = defined $opts->{'sample_rate'} ? $opts->{'sample_rate'} : 1;
    if ($sample_rate == 1 || rand() <= $sample_rate) {
        $stat =~ s/::/./g;
        $stat =~ s/[:|@]/_/g;
        my $rate = '';
        $rate = "|\@${sample_rate}" unless $sample_rate == 1;
        my $tags = '';
        $tags = "|#" . join(',', @{$opts->{'tags'}}) if $opts->{'tags'};
        my $message = $self->{'namespace'} . "${stat}:${delta}|${type}${rate}${tags}";
        return $self->send_to_socket($message);
    }
}

sub send_to_socket {
    my ($self, $message) = @_;

    my $r = send($self->_socket(), $message, 0);
    if (!defined $r) {
        return 0;
    } elsif ($r != length($message)) {
        return 0;
    }

    return 1;
}

1;
__END__

=encoding utf-8

=head1 NAME

DataDog::DogStatsd - A Perl client for DogStatsd

=head1 SYNOPSIS

	use DataDog::DogStatsd;

	my $statsd = DataDog::DogStatsd->new;
	$statsd->increment( 'user.login', { tags => [ $s->cs->param( 'loginName' ) ] } );
	$statsd->timing('page.load', 320);
	$statsd->gauge('users.online', 100);

	# namespace version
	my $statsd = DataDog::DogStatsd->new(namespace => 'account.'); # note the '.' at the ending
	$statsd->increment('activate'); # actually as account.activate

=head1 DESCRIPTION

Statsd: A DogStatsd client (https://www.datadoghq.com)

The difference between L<Net::Statsd> and this is that it supports tags and namespace.

=head1 METHODS

=head2 new

=over 4

=item * host

default to 127.0.0.1

=item * port

default to 8125

=item * namespace

default to ''

=back

=head2 increment

	$statsd->increment( 'page.views' );
	$statsd->increment( 'test.stats', { sample_rate => 0.5 } );
	$statsd->increment( 'user.login', { tags => [ $s->cs->param( 'loginName' ) ] } );

Sends an increment (count = 1) for the given stat to the statsd server.

=head2 decrement

	$statsd->decrement( 'products.available' );

Sends a decrement (count = -1) for the given stat to the statsd server.

=head2 count

	$statsd->count( 'test.stats', 2 );

Sends an arbitrary count for the given stat to the statsd server.

=head2 gauge

	$statsd->gauge('users.online', 100);
	$statsd->gauge('users.online', 100, { tags => ['tag1', 'tag2'] });

Sends an arbitrary gauge value for the given stat to the statsd server.

This is useful for recording things like available disk space, memory usage, and the like, which have different semantics than counters.

=head2 histogram

	$statsd->histogram($stats, $value);

Sends a value to be tracked as a histogram to the statsd server.

=head2 timing

	$statsd->timing('page.load', 320);
	$statsd->timing('page.load', 320, { sample_rate => 0.5, tags => ['tag1', 'tag2'] });

Sends a timing (in ms) for the given stat to the statsd server. The sample_rate determines what percentage of the time this report is sent. The statsd server then uses the sample_rate to correctly track the average timing for the stat.

=head2 set

	$statsd->set('visitors.uniques', $user_id);

Sends a value to be tracked as a set to the statsd server.

=head2 event

	$statsd->event('event title', 'event text', { tags => ['tag1, 'tag2'] });
	$statsd->event('event title', 'event text', { tags => ['tag1', 'tag2'], alert_type => 'error'});

Sends an event indicating a specific incident occurring.

Use alert_type option to specify the kind of event, available options are error, warning, info, and success.

=head1 AUTHORS

Stefan Goethals <stefan@zipkid.eu>
